File: omnipod-dash/src/main/java/info/nightscout/androidaps/plugins/pump/omnipod/dash/driver/pod/command/base/HeaderEnabledCommand.java
package info.nightscout.androidaps.plugins.pump.omnipod.dash.driver.pod.command.base;

import java.nio.ByteBuffer;

import info.nightscout.androidaps.plugins.pump.omnipod.dash.driver.pod.util.MessageUtil;

public abstract class HeaderEnabledCommand implements Command {
    protected static final short HEADER_LENGTH = 6;

    protected final CommandType commandType;
    protected final int uniqueId;
    protected final short sequenceNumber;
    protected final boolean multiCommandFlag;

    protected HeaderEnabledCommand(CommandType commandType, int uniqueId, short sequenceNumber, boolean multiCommandFlag) {
        this.commandType = commandType;
        this.uniqueId = uniqueId;
        this.sequenceNumber = sequenceNumber;
        this.multiCommandFlag = multiCommandFlag;
    }

    @Override public CommandType getCommandType() {
        return commandType;
    }

    protected static byte[] appendCrc(byte[] command) {
        return ByteBuffer.allocate(command.length + 2) //
                .put(command) //
                .putShort(MessageUtil.createCrc(command)) //
                .array();
    }

    protected static byte[] encodeHeader(int uniqueId, short sequenceNumber, short length, boolean multiCommandFlag) {
        return ByteBuffer.allocate(6) //
                .putInt(uniqueId) //
                .putShort((short) (((sequenceNumber & 0x0f) << 10) | length | ((multiCommandFlag ? 1 : 0) << 15))) //
                .array();
    }

}

File: pump/omnipod-dash/src/main/kotlin/app/aaps/pump/omnipod/dash/driver/pod/command/base/HeaderEnabledCommand.kt
package app.aaps.pump.omnipod.dash.driver.pod.command.base

import app.aaps.pump.omnipod.dash.driver.pod.util.MessageUtil
import java.nio.ByteBuffer

abstract class HeaderEnabledCommand protected constructor(
    override val commandType: CommandType,
    protected val uniqueId: Int,
    override val sequenceNumber: Short,
    protected val multiCommandFlag: Boolean
) : Command {

    companion object {

        internal fun appendCrc(command: ByteArray): ByteArray =
            ByteBuffer.allocate(command.size + 2)
                .put(command)
                .putShort(MessageUtil.createCrc(command))
                .array()

        internal fun encodeHeader(
            uniqueId: Int,
            sequenceNumber: Short,
            length: Short,
            multiCommandFlag: Boolean
        ): ByteArray =
            ByteBuffer.allocate(6)
                .putInt(uniqueId)
                .putShort((sequenceNumber.toInt() and 0x0f shl 10 or length.toInt() or ((if (multiCommandFlag) 1 else 0) shl 15)).toShort())
                .array()

        internal const val HEADER_LENGTH: Short = 6
    }
}
